include/odp/api/spec/comp.h

```diff
@@ -0,0 +1,574 @@
+
+	/** Input data range to process */
+	odp_packet_data_range_t in_data_range;
+
+	/** Output packet data range.
+	 * Indicates where processed packet will be written
+	 * where length specifies, data available for output.
+	 */
+	odp_packet_data_range_t out_data_range;
+} odp_comp_packet_op_param_t;
+
+/**
+ * Query compression capabilities
+ *
+ * Output compression capabilities on success.
+ *
+ * @param[out] capa   Pointer to capability structure for output
+ *
+ * @retval 0 on success
+ * @retval <0 on failure
+ */
+int odp_comp_capability(odp_comp_capability_t *capa);
+
+/**
+ * Query supported compression algorithm capabilities
+ *
+ * Output algorithm capabilities.
+ *
+ * @param	comp	Compression algorithm
+ * @param[out]	capa	Compression algorithm capability
+ *
+ * @retval 0 on success
+ * @retval <0 on failure
+ */
+int odp_comp_alg_capability(odp_comp_alg_t comp,
+			    odp_comp_alg_capability_t capa);
+
+/**
+ * Query supported hash algorithm capabilities
+ *
+ * Outputs all supported configuration options for the algorithm.
+ *
+ * @param	hash     Hash algorithm
+ * @param	capa     Hash algorithm capability
+ *
+ * @retval 0 on success
+ * @retval <0 on failure
+ */
+int odp_comp_hash_alg_capability(odp_comp_hash_alg_t hash,
+				 odp_comp_hash_alg_capability_t capa);
+
+/**
+ * Initialize compression session parameters
+ *
+ * Initialize an odp_comp_session_param_t to its default values for
+ * all fields.
+ *
+ * @param param   Pointer to odp_comp_session_param_t to be initialized
+ */
+void odp_comp_session_param_init(odp_comp_session_param_t *param);
+
+/**
+ * Compression session creation
+ *
+ * Create a comp session according to the session parameters. Use
+ * odp_comp_session_param_init() to initialize parameters into their
+ * default values.
+ *
+ * @param	param	Session parameters
+ *
+ * @retval		Comp session handle
+ * @retval		ODP_COMP_SESSION_INVALID on failure
+ */
+odp_comp_session_t
+odp_comp_session_create(const odp_comp_session_param_t *param);
+
+/**
+ * Compression session destroy
+ *
+ * Destroy an unused session. Result is undefined if session is being used
+ * (i.e. asynchronous operation is in progress).
+ *
+ * @param	session		Session handle
+ *
+ * @retval	0 on success
+ * @retval	<0 on failure
+ */
+int odp_comp_session_destroy(odp_comp_session_t session);
+
+/**
+ * Synchronous packet compression operation
+ *
+ * This operation does packet compression in synchronous mode. A successful operation
+ * returns the number of successfully processed input packets and updates the
+ * results in the corresponding output packets. Outputted packets contain
+ * compression results metadata (odp_comp_packet_result_t), which should be
+ * checked for operation status.
+ *
+ * For every input packet entry in 'pkt_in' array, application should pass
+ * corresponding valid output packet handle. If any error occurs during
+ * processing of packets, the API returns with number of entries successfully
+ * processed.
```

Comment:
Assuming 5 packets are processed, each with a different session, why failing to process 1st packet should trigger the failure for all? I mean, we have a 'result' per packet. Why not returning retval 4 and let application to get result per packet with odp_comp_result()?

Reply:
There is a possibility of co-processor queue being full and the input packet was not even sent to the compression coprocessor, In that scenario, it might be better to just return the number of input packets processed, else implementation will have to update all the output packets even when they were not processed and the application will have to check the result for each and every output packet.

Reply:
OK. Makes sense.